File: lua/tst/tiles.lua
require 'pico.check'

pico.init(true)

-- helper: window-pixel mouse set (NW anchor)
local function mouse_w (x, y)
    local prev = pico.set.layer("window")
    pico.set.mouse({'!', x=x, y=y, anchor='NW'})
    pico.set.layer(prev)
end

-- 4x4 grid of 4x4 pixel tiles = 16x16 logical world
-- 160x160 physical window
local phy  = {'!', w=160, h=160}
local log  = {'#', w=4, h=4}
local tile = {w=4, h=4}
pico.set.window { title="Tiles" }
pico.set.layer("window")
pico.set.scene { dim=phy }
pico.set.layer("world")
pico.set.scene { dim=log, tile=tile }

-- 4x4 pixel white tile
local white = {}
for y = 1, 4 do
    white[y] = {}
    for x = 1, 4 do
        white[y][x] = { r=0xFF, g=0xFF, b=0xFF, a=0xFF }
    end
end

-- Test 1: tile (1,1) with NW anchor - top-left corner
do
    print("tile (1,1) NW anchor")
    pico.output.clear()
    local r = {'#', x=1, y=1, w=1, h=1, anchor='NW'}
    pico.output.draw.pixmap("buf", white, r)
    pico.check("tiles-01")
end

-- Test 2: tile (2,2) with C anchor - centered in tile
do
    print("tile (2,2) C anchor")
    pico.output.clear()
    local r = {'#', x=2, y=2, w=1, h=1, anchor='C'}
    pico.output.draw.pixmap("buf", white, r)
    pico.check("tiles-02")
end

-- Test 3: 2x2 tiles at (1,1) with NW anchor
do
    print("2x2 tiles (1,1) NW anchor")
    pico.output.clear()
    local r = {'#', x=1, y=1, w=2, h=2, anchor='NW'}
    pico.output.draw.pixmap("buf", white, r)
    pico.check("tiles-03")
end

-- Test 4: 2x2 tiles centered at (2.5, 2.5)
do
    print("2x2 tiles (2.5,2.5) C anchor")
    pico.output.clear()
    local r = {'#', x=2.5, y=2.5, w=2, h=2, anchor='C'}
    pico.output.draw.pixmap("buf", white, r)
    pico.check("tiles-04")
end

-- Test 5: mouse position in tile mode
-- helper: each call site is exercised in both forms
--   * table form `{'#', anchor='NW'}` — preserves legacy NW numbers
--   * mode-string form — new C-anchor default; result shifted by -0.5
--     per axis in tile units (for '#': `flt/tile + 1 - anchor.x`)

-- phy (0,0) -> log (0,0) -> tile (1,1)
do
    print("mouse tile (1,1)")
    mouse_w(0, 0)
    local pos = pico.get.mouse({'#', anchor='NW'})
    assert(pos.x==1 and pos.y==1)
    local pos = pico.get.mouse(nil, {'#', anchor='NW'})
    assert(pos.x==1 and pos.y==1)
    local pos = pico.get.mouse('#')
    assert(pos.anchor.x==0.5 and pos.anchor.y==0.5)
    assert(pos.x==0.5 and pos.y==0.5)
end

-- phy (40,40) -> log (4,4) -> tile (2,2)
do
    print("mouse tile (2,2)")
    mouse_w(40, 40)
    local pos = pico.get.mouse({'#', anchor='NW'})
    assert(pos.x==2 and pos.y==2)
    local pos = pico.get.mouse('#')
    assert(pos.x==1.5 and pos.y==1.5)
end

-- phy (80,120) -> log (8,12) -> tile (3,4)
do
    print("mouse tile (3,4)")
    mouse_w(80, 120)
    local pos = pico.get.mouse({'#', anchor='NW'})
    assert(pos.x==3 and pos.y==4)
    local pos = pico.get.mouse('#')
    assert(pos.x==2.5 and pos.y==3.5)
end

pico.init(false)

File: tst/tiles.c
#include "pico.h"
#include "../check.h"

// helper: window-pixel mouse set (NW anchor)
static void mouse_w (int x, int y) {
    const char* old = pico_set_layer("window");
    pico_set_mouse(&(Pico_Rel_Pos){'!', {x, y}, PICO_ANCHOR_NW});
    pico_set_layer(old);
}

int main (void) {
    pico_init(1);

    // 4x4 grid of 4x4 pixel tiles = 16x16 logical world
    Pico_Rel_Dim log  = { '#', {4, 4} };
    Pico_Abs_Dim tile = { 4, 4 };
    pico_set_window((Pico_Window){ .fs=0, .show=1, .title="Tile" });

    pico_set_layer("window");
    pico_set_scene_dim(&(Pico_Rel_Dim){ '!', {160, 160} });
    pico_set_layer("world");
    pico_set_scene_tile(tile); pico_set_scene_dim(&log);

    // 4x4 pixel white tile
    Pico_Color white[16];
    for (int i = 0; i < 16; i++) {
        white[i] = (Pico_Color){ 0xFF, 0xFF, 0xFF, 0xFF };
    }

    // Test 1: tile (1,1) with NW anchor - top-left corner
    {
        puts("tile (1,1) NW anchor");
        pico_output_clear();
        Pico_Rel_Rect r = { '#', {1, 1, 1, 1}, PICO_ANCHOR_NW };
        pico_output_draw_pixmap("tile", (Pico_Abs_Dim){4,4}, white, &r);
        _pico_check("tiles-01");
    }

    // Test 2: tile (2,2) with C anchor - centered in tile
    {
        puts("tile (2,2) C anchor");
        pico_output_clear();
        Pico_Rel_Rect r = { '#', {2, 2, 1, 1}, PICO_ANCHOR_C };
        pico_output_draw_pixmap("tile", (Pico_Abs_Dim){4,4}, white, &r);
        _pico_check("tiles-02");
    }

    // Test 3: 2x2 tiles at (1,1) with NW anchor
    {
        puts("2x2 tiles (1,1) NW anchor");
        pico_output_clear();
        Pico_Rel_Rect r = { '#', {1, 1, 2, 2}, PICO_ANCHOR_NW };
        pico_output_draw_pixmap("tile", (Pico_Abs_Dim){4,4}, white, &r);
        _pico_check("tiles-03");
    }

    // Test 4: 2x2 tiles centered at (2.5, 2.5)
    {
        puts("2x2 tiles (2.5,2.5) C anchor");
        pico_output_clear();
        Pico_Rel_Rect r = { '#', {2.5, 2.5, 2, 2}, PICO_ANCHOR_C };
        pico_output_draw_pixmap("tile", (Pico_Abs_Dim){4,4}, white, &r);
        _pico_check("tiles-04");
    }

    // Test 5: mouse position in tile mode
    // phy (0,0) -> log (0,0) -> tile (1,1)
    {
        puts("mouse tile (1,1)");
        mouse_w(0, 0);
        Pico_Mouse pos = pico_get_mouse('#', NULL);
        assert(pos.x==1 && pos.y==1);
    }

    // phy (40,40) -> log (4,4) -> tile (2,2)
    // 160 phy / 16 log = 10x scale, so phy 40 = log 4
    {
        puts("mouse tile (2,2)");
        mouse_w(40, 40);
        Pico_Mouse pos = pico_get_mouse('#', NULL);
        assert(pos.x==2 && pos.y==2);
    }

    // phy (80,120) -> log (8,12) -> tile (3,4)
    {
        puts("mouse tile (3,4)");
        mouse_w(80, 120);
        Pico_Mouse pos = pico_get_mouse('#', NULL);
        assert(pos.x==3 && pos.y==4);
    }

    pico_init(0);
    return 0;
}
